### DISCUSSION TITLE
First programming language

### DISCUSSION THREAD

**Discussion by Borislav-Test** [community/community#27660] [Other Feature Feedback, Questions, & Ideas]
I have been wondering on what programming language to choose as first.

**Comment by BorislavChernev** [comment]
C# Is an easy one. I think its a good language to start with.

**Reply by mosmo0220** [reply]
I will agree with this, being in 6th grade of primary school I was able to write a pretty cool calculator in winforms, which I wouldn't have done in e.g. c++

**Comment by jjaaster** [comment]
I would also suggest HTML and CSS, if they make center stage, and extensional JS if your interested in Web Development.

**Reply by MelphinDev** [reply]
HTML and CSS are not programming languages.

**Reply by Sanjeev-Kumar78** [reply]
Yes,
*HTML is markup languages & CSS is style sheet language.*

**Comment by thaian161** [comment]
> I have been wondering on what programming language to choose as first.

I would go with JS first, it also depends on your purposes tho, would you like to become a Web Dev? Data Analyst? or mobile dev? 
if you want to be a web dev, then go with JS, and probably learn the basics of HTML and CSS as well. 
If you want to do data stuff, learn Python and SQL

check out this video: https://www.youtube.com/watch?v=poJfwre2PIs

Hope this help.

**Comment by redegg89** [comment]
It all depends on what you want to do. 

If you're looking to do things in a web browser, through a tool like Electron or React, JS is for you. Python is the most different in terms of syntax, for better or worse, but there are enough tools for it that it's still worth using, even if not for data science/ai. If you're looking for a language for game development, it really depends on the engine you want to use (or none at all). The two I can think of right now are that [Unity](https://unity.com/) uses C# and [Unreal](https://www.unrealengine.com/en-US) uses C++. Like I mentioned before, you can also use Python for game development with something like [RenPy](https://www.renpy.org/) or [Pygame](https://www.pygame.org/). If you plan to do modding for Minecraft Java, your only option is Java (obviously). Luckily, it doesn't really matter which language you use from my experience, since (aside from Assembly and maybe a few others) you can compile/run your program on whatever you want if you have the right build tools and you can do most things, no matter the language from my (very limited) experience.

**Comment by radar-blip** [comment]
Python

**Comment by Stratis-Dermanoutsos** [comment]
I'd say C#.

OOP is extremely important for any software developer nowadays and the best languages to teach you that is C# and Java.
From these two, C# is the easiest and currently more supported by its developers as Microsoft is seriously optimizing and improving it by the day.

As for JavaScript and Python, yes, they are both very easy to use and beginner-friendly in general.
However, they can make the life of a beginner difficult.

- Javascript has too unique syntax and features for some and does not help with understanding types, scope, inheritance etc.
- Python is missing many OOP features and types as well. Although it is easy to understand and use, beginners miss on too much basic information.

**EDIT**

Of course, if you want to become a web developer you'll need JavaScript and if you like other scientific fields, python is a must. But it'll be better if you have basic and generalized knowledge before moving to something specific and C# is very good for that.

**Comment by DG12** [comment]
I've always felt that the better choice between languages is the one with a very good debugger. This lets you see information the way the language does, not the way you think it it.

**Comment by s4shantanu** [comment]
I say Java,
Because, If we start with the Independent and object orientated then we can quickly shift into any other language like python c# because both are easy also python has many built-in libraries so easy to use also we are clear with our aim to choose a correct domain.

**Comment by Vastyen** [comment]
Python.

After learning Python and how the communication with the computer works, I think the next it's the different paradigms. 
If you can understand the most used paradigms probably with some hours you can domine a lot of programming languages.

So, learn Python and then, functional paradigm, oriented object programming, imperative paradigm and oriented events paradigm.

**Comment by librity** [comment]
Python or Ruby.
I prefer Ruby.

**Reply by s4shantanu** [reply]
I think Python 
This language is easy so newbie can't hate the programming they think is easy to understand python have multiple domains after learning python we have various options in numerous fields.

**Reply by librity** [reply]
> I think Python This language is easy so newbie can't hate the programming they think is easy to understand python have multiple domains after learning python we have various options in numerous fields.

I disagree and I won't further elaborate.

**Reply by s4shantanu** [reply]
Ok its fine,  This is just a my thinking with python and ruby.

**Comment by Robin270** [comment]
I'd recommand to start with JavaScript. For beginners, it could be more attractive, because of the connection with HTML and CSS. For example, in Python, you need to learn about modules etc. to even get some other output than print() to terminal, but in JavaScript, an interesting visual output in the form of modifying a HTML page is one of the first things you can do in this promgramming language. And in general, I also find it a bit easier to understand for beginners than Python, but it still isn't a bad option. That's just my opinion, but it comes from a personal experience when teaching students the very basics of programming. One language I wouldn't recommand for novices at all is the C# – its logic and syntax differs a lot from Python or JavaScript – at least when you're not an experienced programmer. But if none of these two I said are fine to start with, I can also imagine starting on PHP – IMHO very similar to JavaScript and much more powerful (since it's a back-end programming language) and you can do those powerful things right from the start, which is the main difference from Python which is also a back-end programming language.

**Comment by happy-spark** [comment]
scratch
XD

**Comment by librity** [comment]
Any one that sparks your interest, they're all pretty much the same.

**Comment by MelphinDev** [comment]
I don't recommend JavaScript. It has so many odd behaviours.
I recommend Python. It's somewhat easy to learn.

**Comment by DG12** [comment]
Python diagnostics are terrible. Syntax error. OK on line xxx but give me a hit :-(

Where's the debugger? The best language is one that shows what IT thinks the contents of variables are by displaying them in a debugger.

I'm tending to go with javascript. Browsers include great debugging. You can use and share what you write.

**Comment by Rinky6767** [comment]
> I have been wondering on what programming language to choose as first.

I will suggest you to first start of with C language and then move on to C++. Once you are done with C and C++ you will have solid foundation in programming because this languages are very good to understand procedural as well as object oriented way of programming.
All the best👍 
Hope this answers your question:)

**Reply by librity** [reply]
I think C then Rust if you wanna be a Software Engineer.
C++ is outdated and unwieldy, and it's a terrible introduction to OOP.
If you just wanna learn how to write some code and automate your workflow ruby and python are perfectly acceptable languages.

**Reply by Rinky6767** [reply]
It's not about being outdated talking about clearing the basic concept.
Sorry but C++ and also Java is very good for beginners to start with object oriented programming.
Even Java is very good for Data structures and Algorithms.

**Reply by airtower-luna** [reply]
The problem with C and C++ is that they don't ensure safe memory access. This is a problem even for experienced programmers (many, many security issues come from that!), so I wouldn't recommend beginners start with either unless they really need the specific language. Rust is an excellent starting point if you want something just as close to hardware, or maybe Go.

For a higher abstraction level Java is a decent choice if you want a curly brace language, but if you don't care about the braces I prefer Python.

**Reply by librity** [reply]
C is what you wanna learn if you're gonna be a software engineer.
You're not a software engineer until you've learned to manage memory.

**Comment by AyushJain001** [comment]
I learned java first that in my school days and I think that was good because after that other languages looks so easy and concepts are same though . After java I enhanced in Python which I needed for my career in data science and machine learning domain , for DSA I practice both in python and C++ as cpp provide more growth wrt to STL libraries

**Comment by HannaPleshko** [comment]
JavaScript

**Comment by ChinoBeach-Colgate** [comment]
I started with Python and then worked my way to the C languages and Java, but it really depends what you want to do. If you want to make games C++ or C# are the best bet.

**Reply by ChinoBeach-Colgate** [reply]
additionally I also recommend taking an online course with examples to help. Going along the game dev/programming line id recommend something like GameDevTv. They often have courses on sale on either their own site or through humble bundle

**Comment by celebrateseo** [comment]
Can anyone rank each of the relevant programming languages in ranking order for me and for beginners to see which one matters most and as we ahead, which one are the essentials. Its also part of my research for a future post at **[celebrity wishes](https://celedaily.com)**

**Comment by sh0rtener** [comment]
I think that C# it's not bad variant to start programming. Starting with syntax sugar and ending with universality of the language, it'll be possible to understanding what do you want to develop as developer and which direction you want to choose and also will know a basis of the programming such as: OOP, structures and e.t.c

**Comment by DevModee** [comment]
java
